walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,86 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> availableTimesForMandatoryAttendeesOnly =
+      getAvailableTimes(events, request, request.getAttendees());
+    ArrayList<TimeRange> availableTimesForOptionalAttendeesOnly = 
+      getAvailableTimes(events, request, request.getOptionalAttendees());
+    
+    ArrayList<TimeRange> availableTimesForAllAttendees = 
+      new ArrayList<TimeRange>(availableTimesForMandatoryAttendeesOnly);
+    availableTimesForAllAttendees.retainAll(availableTimesForOptionalAttendeesOnly);
+
+    if(availableTimesForAllAttendees.size() > 0) return availableTimesForAllAttendees;
+
+    if(request.getAttendees().size() == 0 && request.getOptionalAttendees().size() > 0) {
+      return availableTimesForOptionalAttendeesOnly;
+    }
+
+    return availableTimesForMandatoryAttendeesOnly;
+    
+  }
+
+
+  private ArrayList<TimeRange> getAvailableTimes(Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) return new ArrayList<TimeRange>();
+
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    availableTimes.add(TimeRange.WHOLE_DAY);
+
+    for(Event event : events) {
+
+      boolean eventHasMeetingAttendee = false;
+      for(String eventAttendee : event.getAttendees()) {
+        if(attendees.contains(eventAttendee)) {
+          eventHasMeetingAttendee = true;
+          break;
+        }
+      }
+      if(!eventHasMeetingAttendee) continue;
+
+      TimeRange eventTime = event.getWhen();
+      for(int i = 0; i < availableTimes.size(); i++) {
```

Comment:
I think with the way this loop works you're always looking at the top element since ```i``` never decreases and ```availableTimes``` only ever increases by a size of 1. So using a for loop can be a little more confusing here especially since most times a for loop is used you assume that the size of the collection is fixed. I think a better approach might be maintaining a stack for the ```availableTimes```, and using some kind of while loop. That would better communicate the implementation in my opinion.